### PR TITLE
Fixed #25068 -- Added metaclassmaker to resolve metaclass conflicts in migrations state.

### PR DESCRIPTION
**Pull request by kosz85**
https://github.com/django/django/pull/4968

Now it's faster, or more like it's not slowing down ;)

### CODE REVIEW

**Comment by timgraham** [comment]
@MarkusH - could you run your benchmarks on this or share how we can do it? I'll happily review the patch for style, etc. if it passes that stage of review.

**Comment by MarkusH** [comment]
I just use https://github.com/MarkusH/django-migrations-benchmark as some kind of benchmarking tool and uncomment the code in Django I don't want to have affect on the timing.

**Comment by kosz85** [comment]
@MarkusH I used your benchmark, it seems my current solution doesn't slow down migrations. I have two thoughts after using your benchmark :) Please add `__init__.py` to your benchmark apps :) It will be easier to use by everyone that need to test it. And second one: this doesn't test my case, so after merging it would be nice to improve it and test also models with multiple metadatas.

Here are results, I ran it when working as I don't have spare computer so it's not best measurement, but I think is enough. As you can see it's almost the same, even better with fix, but difference is caused by IO.
```
With fix
real	7m7.687s
user	3m4.366s
sys	0m3.366s

real	7m15.488s
user	3m4.684s
sys	0m3.174s

Without fix
real	7m36.196s
user	3m16.878s
sys	0m3.293s

real	7m24.973s
user	3m11.605s
sys	0m3.414s
```

File: tests/migrations/test_state.py [hunk 1]
```diff
@@ -870,6 +871,37 @@ class Meta:
...
+            pass
+
+        class Mixin(six.with_metaclass(DummyMetaClass, object)):
+            mixin_attr = True
+
+        class ItermidiateMetaClass(ModelBase, DummyMetaClass):
+            pass
+
+        # Create the same Model without metaclassmaker as you would normally do.
+        class MetaclassTestModel(six.with_metaclass(ItermidiateMetaClass, models.Model, Mixin)):
```
Comment (MarkusH) [inline tests/migrations/test_state.py]: `Mixin` should come before `models.Model`

File: tests/migrations/test_state.py [hunk 2]
```diff
@@ -870,6 +871,37 @@ class Meta:
...
+        """
+        new_apps = Apps(['migrations'])
+
+        class DummyMetaClass(type):
+            pass
+
+        class Mixin(six.with_metaclass(DummyMetaClass, object)):
+            mixin_attr = True
+
+        class ItermidiateMetaClass(ModelBase, DummyMetaClass):
```
Comment (MarkusH) [inline tests/migrations/test_state.py]: Can you swap `ModelBase` and `DummyMetaClass` here, please.

File: tests/migrations/test_state.py [hunk 3]
```diff
@@ -870,6 +871,37 @@ class Meta:
...
+            pass
+
+        # Create the same Model without metaclassmaker as you would normally do.
+        class MetaclassTestModel(six.with_metaclass(ItermidiateMetaClass, models.Model, Mixin)):
+            class Meta:
+                app_label = "migrations"
+                apps = new_apps
+                swappable = 'TEST_SWAPPABLE_MODEL'
```
Comment (MarkusH) [inline tests/migrations/test_state.py]: I don't think we should have that here

**Comment by MarkusH** [comment]
Can you make `ItermidiateMetaClass` and `DummyMetaClass` each add an attribute the object, please.

**Comment by kosz85** [comment]
`ItermidiateMetaClass` shouldn't add special attr, I can add it to DummyMetaClass thought.

File: tests/migrations/test_state.py (hunk 1 above)
Comment (kosz85) [inline tests/migrations/test_state.py]: I will change it, but you know that this doesn't really matter here? ;) It's just for you.

File: tests/migrations/test_state.py [hunk 4]
```diff
@@ -870,6 +871,38 @@ class Meta:
...
 
+    def test_metaclass_conflict(self):
+        """
+        Test making class with complex metadata inheritance
+        """
+        new_apps = Apps(['migrations'])
+
+        class DummyMetaClass(type):
+            dummy_attr = 'dummy'
+            pass
```
Comment (MarkusH) [inline tests/migrations/test_state.py]: You don't need this pass anymore.

Comment (kosz85) [inline tests/migrations/test_state.py]: hehe, you are right :) You want to tire me with this topic to death ;) will fix in minutes.

**Comment by MarkusH** [comment]
@timgraham if you want to merge, patch looks good to me.

**Comment by MarkusH** [comment]
Close in favor of #5334

**Pull request closed without merging**